plugin/externalPluginEntry.go

```diff
@@ -138,76 +140,42 @@ func (e *externalPluginEntry) Metadata(ctx context.Context) (JSONObject, error)
 			strings.TrimSpace(string(stdout)),
 			"{\"key1\":\"value1\",\"key2\":\"value2\"}",
 		)
-
 		return nil, fmt.Errorf("could not decode the metadata from stdout: %v", err)
 	}
-
 	return metadata, nil
 }
 
-type stdoutStreamer struct {
-	stdoutRdr io.ReadCloser
-	cmd       *exec.Cmd
-}
-
-func (s *stdoutStreamer) Read(p []byte) (int, error) {
-	return s.stdoutRdr.Read(p)
-}
-
-func (s *stdoutStreamer) Close() error {
-	var err error
-
-	if closeErr := s.stdoutRdr.Close(); closeErr != nil {
-		err = fmt.Errorf("error closing stdout: %v", closeErr)
-	}
-
-	if signalErr := s.cmd.Process.Signal(syscall.SIGTERM); signalErr != nil {
-		signalErr = fmt.Errorf(
-			"error sending SIGTERM to process with PID %v: %v",
-			s.cmd.Process.Pid,
-			signalErr,
-		)
-
-		if err != nil {
-			err = fmt.Errorf("%v; and %v", err, signalErr)
-		} else {
-			err = signalErr
-		}
-	}
-
-	return err
-}
-
-// Stream streams the entry's content
 func (e *externalPluginEntry) Stream(ctx context.Context) (io.ReadCloser, error) {
-	cmd, stdoutR, stderrR, err := CreateCommand(
-		e.script.Path(),
-		"stream",
-		e.id(),
-		e.state,
-	)
-
+	cmd := e.script.NewInvocation(ctx, "stream", e)
+	stdoutR, err := cmd.StdoutPipe()
 	if err != nil {
 		return nil, err
 	}
-
-	cmdStr := fmt.Sprintf("%v %v %v %v", e.script.Path(), "stream", e.id(), e.state)
-
-	activity.Record(ctx, "Starting command: %v", cmdStr)
+	stderrR, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, err
+	}
+	activity.Record(ctx, "Starting %v", cmd)
 	if err := cmd.Start(); err != nil {
-		activity.Record(ctx, "Closed command stdout: %v", stdoutR.Close())
-		activity.Record(ctx, "Closed command stderr: %v", stderrR.Close())
 		return nil, err
 	}
+	// "wait" will be used in Stream's error handlers. It will be wrapped
+	// in a "defer" call to ensure that cmd.Wait()'s called once we've read
```

Comment:
My reason for doing it here is to enforce the idea that "If Stream outputs something to stderr, something went wrong w/ it so the process should have terminated. We can check the latter by calling Wait(). If something went wrong with the termination, or if the process was killed by the context-cancellation cleanup, we can log Wait's error." For stdout, I moved the `Wait()` over to `Close` because that's when we want to wait for termination and release the resources. So I think this code here is correct.

Reply:
Ok.